### PR TITLE
feat: ✨ add support for reading client credentials from the local H2O CLI config file

### PR DESCRIPTION
**Pull request by zoido**
- implement config path lookup
 - implement config loading
 - extend credentials loading to accept tokens loaded from the config
 - extend main discover functions to accept config path and utilize

RESOLVES https://github.com/h2oai/cloud-discovery/issues/437

### CODE REVIEW

**Review by geomodular — APPROVED** [review]
Thanks @zoido! LGTM. I'm rusty in Python. I haven't touched it for a while.

File: src/h2o_discovery/_internal/config.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+def _empty_tokens_factory() -> Mapping[str, str]:
+    return types.MappingProxyType({})
+
+
+@dataclasses.dataclass(frozen=True)
+class Config:
+    """Internal representation of the H2O CLI Configuration."""
+
+    #: Map of found tokens in the configuration file. in the `{"client-id": "token"}`
```
[the remark this answers is not in the document]
Comment (zoido) [reply src/h2o_discovery/_internal/config.py:17]: These are Sphinx comment.
Comments like these are recognized by the documentation generator as field doc strings.

There are some alternatives how to do these but we fell back to this. Discussion was held in the scope of https://github.com/h2oai/cloud-discovery-py/pull/48.

File: src/h2o_discovery/__init__.py [hunk 2]
```diff
@@ -11,44 +14,79 @@
 
 
 def discover(
-    environment: Optional[str] = None, discovery_address: Optional[str] = None
+    environment: Optional[str] = None,
+    discovery_address: Optional[str] = None,
+    config_path: Optional[Union[str, bytes, os.PathLike]] = None,
```
[the remark this answers is not in the document]
Comment (zoido) [reply src/h2o_discovery/__init__.py:19]: I was thinking about having a type in the config module as that's the place that ultimately consumes it. Then I was thinking that exposed API should not depend on types from internal package.

As it documents the API I lean to less indirect and more verbose side.

File: pyproject.toml [hunk 3]
```diff
@@ -14,7 +14,10 @@ classifiers = [
   "Topic :: Security",
   "Topic :: Utilities",
 ]
-dependencies = ["httpx>=0.16"]
+dependencies = [
+  "httpx>=0.16",
+  "tomli >= 1.1.0 ; python_version < \"3.11\"",
```
[the remark this answers is not in the document]
Comment (zoido) [reply pyproject.toml:19]: Good catch. It needs to be quoted but single quotes are also allowed. Much nicer now without escaping.

ref: https://peps.python.org/pep-0508/

**Pull request merged**